token/jwt/jwt.go

```diff
@@ -25,6 +25,8 @@ func (j *RS256JWTStrategy) Generate(claims jwt.Claims, header Mapper) (string, s
 
 	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
 	token.Header = assign(token.Header, header.ToMap())
+	// "public" is understood to be the most recent keyid
+	token.Header["kid"] = "public"
```

Comment:
It should be enough to add that header field in [`NewSession()`](https://github.com/ory/hydra/blob/master/oauth2/session.go#L21)